### PR TITLE
Clarify b0 threshold

### PR DESCRIPTION
**Pull request by EmmaRenauld**
# Quick description

1. [ENH] The argument --force_b0_threshold, used a bit everywhere, was quite unclear. Does it force to use the b0_threshold given, or does it force to work in any case? Answer was: it forced to use the minimal b-value even if it was higher than expected b0_threshold. Renamed --skip_b0_validation.

2.  [FIX] In many cases, the function `check_b0_threshold` was incorrectly used in the scripts (they were using `check_b0_threshold(args)` but the function expected a bool as first argument. It was thus not doing as expected.

3. [FIX] In other cases, it was used with: `b0_thr = check_b0_threshold(force_b0_threshold, bvals.min(), bvals.min())` which was not doing anything!

4. [ENH] It was always used with the default b0_threshold. Added an argument b0_threshold that user may change if there is an error.

5. [FIX] Major fix in scil_sh_to_sf. With option --in_bval, The minimal b-value was always considered as a b0 and rejected from the average, even if the --in_bval contained no b0, which is actually kind of expected in this particular script.

# Complex description :)

**Easy scripts:**
   - scil_dwi_detect_volume_outliers: Ok. Usage was already clean. Just using the new official arg.
   - scil_dwi_extract_b0: Idem.
   - scil_dwi_to_sh: Ok. Idem. However, arg --force_b0 existed but was not used anywhere! Fixed.

**More complicated ones:**
   - scil_fodf_ssst: The usage of a b0_threshold seems straightforward for me but was not implemented in this script! By using `check_b0_threshold(args.force_b0_threshold, bvals.min(), bvals.min())`, it was always setting the b0_thr to the minimal b-value! But dipy's csdeconv does seem to use the gtab.b0s_mask. Added the b0_threshold option.
   - scil_sh_to_sf: There was no option. Used the minimal b-value as threshold. Option added.
   - scil_qball_metrics: Option existed, but then gtab was created with bvals.min(). Fixed, and added option b0_threshold.
   - scil_frf_ssst: The usage of a b0_threshold seems straightforward for me but was not implemented in this script!  Added the b0_threshold option.

**Impossible ones for now:**
   - scil_dti_metrics: I added some text. The gtab.b0_masks does not seem used anywhere in dipy's tensor model I think. To be confirmed by some expert.
   - scil_fodf_msmt: Cannot manage a separate threshold right now. Added link to dipy's issue. Added explanation. Using tolerance as b0_threshold.
   - scil_dki_metrics: Could not modify the identify_shells method right now. Only using tolerance. Link to dipy's issue also added. But if you think usage is clear in dki, please tell me what it does, I will modify the text.
   - scil_frf_msmt: Could not modify the extract_dwi_shell method right now. Only using tolerance. 
   - scil_frf_memsmt: Could not modify the genereate_btensor_input method right now. Only using tolerance. No checks at all here.
   - scil_fodf_memsmt: Idem.
   - scil_btensor_metrics: Idem. Removed unused option force_b0.

## Type of change

Check the relevant options.

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Provide data, screenshots, command line to test (if relevant)

...

# Checklist

- [ x] My code follows the style guidelines of this project (run [autopep8](https://pypi.org/project/autopep8/))
- [ x] I added relevant citations to scripts, modules and functions docstrings and descriptions
- [ x] I have performed a self-review of my code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] I have made corresponding changes to the documentation
- [ x] My changes generate no new warnings
- [ x] I moved all functions from the script file (except the argparser and main) to scilpy modules
- [ x] I have added tests that prove my fix is effective or that my feature works
- [ x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by EmmaRenauld** [comment]
See associated Issue #866

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-867/1/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-867/2/

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-867/8/

**Comment by EmmaRenauld** [comment]
Should be ready for second review, @arnaudbore , @karanphil 
Hoping we will not change our minds again!

Adding here reference to dipy issue: https://github.com/dipy/dipy/issues/3015

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scilpy/gradients/bvec_bval_tools.py [hunk 1]
```diff
@@ -65,58 +65,62 @@ def normalize_bvecs(bvecs, filename=None):
...
+    b0_thr = args.b0_threshold
     if b0_thr > DEFAULT_B0_THRESHOLD:
         logging.warning(
-            'Warning: Your defined threshold is {}. This is suspicious. We '
-            'recommend using volumes with bvalues no higher '
-            'than {} as b0s.'.format(b0_thr, DEFAULT_B0_THRESHOLD)
-        )
+            'Warning: Your defined b0 threshold (option --{}) is {}. This is '
+            'suspicious. We recommend using volumes with bvalues no higher '
+            'than {} as b0s.'.format(b0_thr, b0_thr, DEFAULT_B0_THRESHOLD))
```
Comment (karanphil) [inline scilpy/gradients/bvec_bval_tools.py:99]: I think the first format should be "b0_threshold".

File: scilpy/gradients/bvec_bval_tools.py [hunk 2]
```diff
@@ -65,58 +65,62 @@ def normalize_bvecs(bvecs, filename=None):
...
 
-    if bvals_min < 0:
+    if min_bval < 0:
         logging.warning(
-            'Warning: Your dataset contains negative b-values (minimal '
-            'bvalue of {}). This is suspicious. We recommend you check '
-            'your data.')
+            'Warning: Your dataset contains negative b-values (minimal bvalue '
+            'of {}). This is suspicious. We recommend you check your data.')
```
Comment (karanphil) [inline scilpy/gradients/bvec_bval_tools.py:104]: Missing a format?

File: scilpy/gradients/bvec_bval_tools.py [hunk 3]
```diff
@@ -65,58 +65,62 @@ def normalize_bvecs(bvecs, filename=None):
...
-                'Warning: Your minimal bvalue is {}, but the threshold '
-                'is set to {}. Since --force_b0_threshold was specified, '
-                'the script will proceed with a threshold of {}'
-                '.'.format(bvals_min, b0_thr, bvals_min))
-            return bvals_min
+                'Warning: Your minimal bvalue ({}), is above the threshold '
+                'defined with --{}: {}.\n'
+                'Since --skip_b0_validation was specified, the script will'
+                'proceed with a b0 threshold of {}.'
+                .format(min_bval, b0_thr, b0_thr, min_bval))
```
Comment (karanphil) [inline scilpy/gradients/bvec_bval_tools.py:114]: I think the second format should be "b0_threshold".

File: scilpy/gradients/bvec_bval_tools.py [hunk 4]
```diff
@@ -65,58 +65,62 @@ def normalize_bvecs(bvecs, filename=None):
...
-                             'regardless.'
-                             .format(bvals_min, b0_thr))
-
+            raise ValueError(
+                'The minimal bvalue ({}) is is above the threshold '
+                'defined with --{}: {}.\n.No b0 volumes can be found.\n'
+                'Please check your data to ensure everything is correct.\n'
+                'You may also increase the threshold or use '
+                '--skip_b0_validation'
+                .format(min_bval, b0_thr, b0_thr, min_bval))
```
Comment (karanphil) [inline scilpy/gradients/bvec_bval_tools.py:123]: I think the first format should be "b0_threshold".

File: scilpy/io/btensor.py [hunk 5]
```diff
@@ -156,7 +157,7 @@ def generate_btensor_input(in_dwis, in_bvals, in_bvecs, in_bdeltas,
         gtab_infos[3] = acq_index_full
         if np.sum([ubvals_full < tol]) < acq_index - 1:
             gtab_infos[3] *= 0
-        return(pa_signals, gtab_infos)
+        return pa_signals, gtab_infos
```
Comment (karanphil) [inline scilpy/io/btensor.py:160]: I would revert back to the parenthesis. This is not the PR to do this :P

File: scilpy/io/btensor.py [hunk 6]
```diff
@@ -185,4 +186,4 @@ def generate_btensor_input(in_dwis, in_bvals, in_bvecs, in_bdeltas,
                                b0_threshold=bvals_full.min(),
                                btens=b_shapes)
 
-    return(gtab_full, data_full, ubvals_full, ub_deltas_full)
+    return gtab_full, data_full, ubvals_full, ub_deltas_full
```
Comment (karanphil) [inline scilpy/io/btensor.py:189]: Same for the parenthesis.

File: scilpy/io/utils.py [hunk 7]
```diff
@@ -217,11 +217,21 @@ def add_overwrite_arg(parser):
         help='Force overwriting of the output files.')
 
 
-def add_force_b0_arg(parser):
-    parser.add_argument('--force_b0_threshold', action='store_true',
-                        help='If set, the script will continue even if the '
-                             'minimum bvalue is suspiciously high ( > {})'
-                        .format(DEFAULT_B0_THRESHOLD))
+def add_skip_b0_validation_arg(parser, b0_tol_name='--b0_threshold'):
```
Comment (karanphil) [inline scilpy/io/utils.py:220]: I think I prefer `add_skip_b0_check_arg`, since the function is called `check_b0_threshold`.

Comment (karanphil) [reply scilpy/io/utils.py:220]: So `--skip_b0_check`. Also shorter to write ;)

File: scilpy/reconst/frf.py [hunk 8]
```diff
@@ -10,12 +10,13 @@
...
 from scilpy.gradients.bvec_bval_tools import (check_b0_threshold,
-                                              is_normalized_bvecs, normalize_bvecs)
+                                              is_normalized_bvecs,
+                                              normalize_bvecs)
 
 
-def compute_ssst_frf(data, bvals, bvecs, mask=None, mask_wm=None,
-                     fa_thresh=0.7, min_fa_thresh=0.5, min_nvox=300,
-                     roi_radii=10, roi_center=None, force_b0_threshold=False):
+def compute_ssst_frf(data, bvals, bvecs, b0_threshold,
```
Comment (karanphil) [inline scilpy/reconst/frf.py:17]: I would put `b0_threshold` optional, with the default value DEFAULT_B0_THRESHOLD.

File: scilpy/reconst/frf.py [hunk 9]
```diff
@@ -221,9 +222,10 @@ def compute_msmt_frf(data, bvals, bvecs, btens=None, data_dti=None,
         logging.warning('Your b-vectors do not seem normalized...')
         bvecs = normalize_bvecs(bvecs)
 
-    b0_thr = check_b0_threshold(force_b0_threshold, bvals.min(), bvals.min())
-
-    gtab = gradient_table(bvals, bvecs, btens=btens, b0_threshold=b0_thr)
+    # toDo. Using a fake b0_threshold here because currently, the
```
Comment (karanphil) [inline scilpy/reconst/frf.py:225]: I would keep the `check_b0_threshold` but with `tol` here. And pass `tol` to the gtab function. Also, maybe remove the toDo? :)

File: scilpy/reconst/sh.py [hunk 10]
```diff
@@ -10,25 +10,27 @@
...
-from scilpy.gradients.bvec_bval_tools import (check_b0_threshold,
-                                              identify_shells,
+from scilpy.gradients.bvec_bval_tools import (identify_shells,
                                               is_normalized_bvecs,
                                               normalize_bvecs)
 from scilpy.dwi.operations import compute_dwi_attenuation
 
 
-def compute_sh_coefficients(dwi, gradient_table, sh_order=4,
+def compute_sh_coefficients(dwi, gradient_table, b0_threshold, sh_order=4,
```
Comment (karanphil) [inline scilpy/reconst/sh.py:19]: Again, make `b0_threshold` optional.

File: scripts/scil_frf_memsmt.py [hunk 11]
```diff
@@ -50,15 +50,13 @@
...
-                             add_overwrite_arg, add_verbose_arg,
+from scilpy.io.utils import (add_overwrite_arg, add_verbose_arg,
                              assert_inputs_exist, assert_outputs_exist,
                              assert_roi_radii_format)
 from scilpy.reconst.frf import compute_msmt_frf
 
 
-def buildArgsParser():
-
+def build_argparser():
```
Comment (arnaudbore) [inline scripts/scil_frf_memsmt.py:59]: Should be _build_arg_parser

File: scripts/scil_btensor_metrics.py [hunk 12]
```diff
@@ -185,14 +184,16 @@ def main():
 
     affine = extract_affine(args.in_dwis)
 
-    tol = args.tolerance
-
+    # Note. This script does not currently allow using a separate b0_threshold
```
Comment (karanphil) [inline scripts/scil_btensor_metrics.py:187]: I think we should add a `check_b0_threshold` using `args.tolerance` here. You would simply have to check if `args.b0_threshold` exists in`check_b0_threshold`, and if not, `b0_thr = args.tolerance` (if it exists).

File: scripts/scil_dwi_detect_volume_outliers.py [hunk 13]
```diff
@@ -67,14 +63,13 @@ def main():
     dwi = nib.load(args.in_dwi)
     data = dwi.get_fdata()
 
-    b0_thr = check_b0_threshold(args.force_b0_threshold,
-                                bvals.min(), args.b0_thr)
+    args.b0_threshold = check_b0_threshold(bvals.min(), args)
```
Comment (karanphil) [inline scripts/scil_dwi_detect_volume_outliers.py:66]: Why putting `args.b0_threshold = ` here?

File: scripts/scil_fodf_memsmt.py [hunk 14]
```diff
@@ -160,17 +160,19 @@ def main():
 
     affine = extract_affine(args.in_dwis)
 
-    tol = args.tolerance
-
     wm_frf = np.loadtxt(args.in_wm_frf)
     gm_frf = np.loadtxt(args.in_gm_frf)
     csf_frf = np.loadtxt(args.in_csf_frf)
 
+    # Note. This script does not currently allow using a separate b0_threshold
```
Comment (karanphil) [inline scripts/scil_fodf_memsmt.py:167]: Same here, I think we should add a `check_b0_threshold` using `args.tolerance`.

File: scripts/scil_fodf_msmt.py [hunk 15]
```diff
@@ -68,14 +69,15 @@ def _build_arg_parser():
              'mask will be used for computations and reconstruction.')
     p.add_argument(
         '--tolerance', type=int, default=20,
-        help='The tolerated gap between the b-values to '
-             'extract and the current b-value. [%(default)s]')
-
-    add_force_b0_arg(p)
+        help='The tolerated gap between the b-values to extract and the '
```
Comment (karanphil) [inline scripts/scil_fodf_msmt.py:72]: I would add this explanation to all `--tolerance` help.

File: scripts/scil_fodf_msmt.py [hunk 16]
```diff
@@ -160,7 +160,16 @@ def main():
...
         bvecs = normalize_bvecs(bvecs)
-    gtab = gradient_table(bvals, bvecs, b0_threshold=b0_thr)
+
+    # Note. This script does not currently allow using a separate b0_threshold
+    # for the b0s. Using the tolerance. To change this, we would have to
+    # change many things in dipy. An issue has been added in dipy to
+    # ask them to clarify the usage of gtab.b0s_mask. See here:
+    #  https://github.com/dipy/dipy/issues/3015
+    # b0_threshold option in gradient_table probably unused.
+    args.b0_threshold = args.tolerance
```
Comment (karanphil) [inline scripts/scil_fodf_msmt.py:170]: I see you did it here, checking the b0 with the tolerance. Might be easier with my proposition using the args. Also, why is there a `args.b0_threshold = `?

File: scripts/scil_frf_memsmt.py [hunk 17]
```diff
@@ -202,17 +199,17 @@ def main():
 
     roi_radii = assert_roi_radii_format(parser)
 
-    tol = args.tolerance
-    dti_lim = args.dti_bval_limit
-    force_b0_thr = args.force_b0_threshold
-
+    # Note. This script does not currently allow using a separate b0_threshold
```
Comment (karanphil) [inline scripts/scil_frf_memsmt.py:202]: Again, I would add a check_b0.

File: scripts/scil_frf_msmt.py [hunk 18]
```diff
@@ -174,14 +174,19 @@ def main():
...
     bvals, bvecs = read_bvals_bvecs(args.in_bval, args.in_bvec)
 
-    tol = args.tolerance
     dti_lim = args.dti_bval_limit
 
-    list_bvals = unique_bvals_tolerance(bvals, tol=tol)
+    # Note. This script does not currently allow using a separate b0_threshold
+    # for the b0s. Using the tolerance. To fix this, we would need to change
+    # the unique_bvals_tolerance and extract_dwi_shell methods.
+    args.b0_threshold = args.tolerance
```
Comment (karanphil) [inline scripts/scil_frf_msmt.py:182]: Might change this part if you wanna use args.tolerance in the function instead.

**Review by karanphil — COMMENTED** [review]
Good job! I've got a few comments, but after they are adressed I think we are gtg.

File: scripts/scil_frf_msmt.py [hunk 19]
```diff
@@ -150,15 +151,14 @@ def buildArgsParser():
...
-    add_force_b0_arg(p)
     add_overwrite_arg(p)
 
     return p
 
 
 def main():
 
-    parser = buildArgsParser()
+    parser = build_argparser()
```
Comment (arnaudbore) [inline scripts/scil_frf_msmt.py:161]: same thing here

File: scripts/scil_fodf_msmt.py (hunk 16 above)
Comment (arnaudbore) [inline scripts/scil_fodf_msmt.py:170]: PK commented this.

**Review by arnaudbore — COMMENTED** [review]
A couple of things but we are getting there !

File: scilpy/reconst/frf.py [hunk 20]
```diff
@@ -152,6 +150,11 @@ def compute_msmt_frf(data, bvals, bvecs, btens=None, data_dti=None,
         1D bvals array with shape (N,)
     bvecs : ndarray
         2D bvecs array with shape (N, 3)
+    btens: ?
```
[the remark this answers is not in the document]
Comment (EmmaRenauld) [reply scilpy/reconst/frf.py:153]: So bvecs_dti, : of shape M, 3.

And btens_dti? What's the difference with btens?

File: scilpy/reconst/frf.py (hunk 9 above)
Comment (EmmaRenauld) [reply scilpy/reconst/frf.py:225]: It's still there but we the way it's implemented right now (passing args), it was easier to do the check in the script than in the function.

File: scilpy/reconst/frf.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (karanphil) [reply scilpy/reconst/frf.py:153]: Its shape M also

File: scripts/scil_dwi_detect_volume_outliers.py (hunk 13 above)
Comment (EmmaRenauld) [reply scripts/scil_dwi_detect_volume_outliers.py:66]: That's the way the --skip_b0 works. If you minimal bvalue is above the tolerance, it returns the minimal b-value instead.

File: scilpy/reconst/frf.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (karanphil) [reply scilpy/reconst/frf.py:153]: All the _dti are just a subset of the other ones, selected for DTI fit.

File: scripts/scil_fodf_memsmt.py (hunk 14 above)
Comment (EmmaRenauld) [reply scripts/scil_fodf_memsmt.py:167]: Ok, but if I do, with the --skip_b0_validation, the args.tolerance value could eventually be changed.

Comment (karanphil) [reply scripts/scil_fodf_memsmt.py:167]: Not sure I understand what you mean here

File: scripts/scil_fodf_msmt.py (hunk 16 above)
Comment (EmmaRenauld) [reply scripts/scil_fodf_msmt.py:170]: Let's talk about it tomorrow. I changed the way to implement it a few times, this seemed best to me, but I'll explain.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-867/12/

**Comment by EmmaRenauld** [comment]
I think most of it is fixed.

I added tests with wrong b0 values. Can't do it with btensor files because in our test data, the minimal b-value is 0.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Review by mdesco — APPROVED** [review]
Looks good to me. Did not test every script. Tested one and read all the --help.

Massive PR. Nice job!

**Pull request merged**